contracts/JBMultiTerminal.sol

```diff
@@ -448,149 +471,247 @@ abstract contract JBPayoutRedemptionPaymentTerminal3_1_2 is
+            _accountingContext.decimals = _accountingContextConfig.standard
+                == JBTokenStandards.NATIVE
+                ? 18
+                : IERC20Metadata(_accountingContextConfig.token).decimals();
+            _accountingContext.currency = uint32(uint160(_accountingContextConfig.token));
+            _accountingContext.standard = _accountingContextConfig.standard;
+
+            // Add the token to the list of accepted tokens of the project.
+            _accountingContextsOf[_projectId].push(_accountingContext);
+
+            emit SetAccountingContext(
+                _projectId, _accountingContextConfig.token, _accountingContext, msg.sender
+            );
+
+            unchecked {
+                ++_i;
+            }
+        }
+    }
+
     //*********************************************************************//
-    // ----------------------- public transactions ----------------------- //
+    // ---------------------- internal transactions ---------------------- //
     //*********************************************************************//
 
+    /// @notice Accepts an incoming token.
+    /// @param _projectId The ID of the project for which the transfer is being accepted.
+    /// @param _token The token being accepted.
+    /// @param _amount The amount of tokens being accepted.
+    /// @param _metadata The metadata in which permit2 context is provided.
+    /// @return The amount of tokens that have been accepted.
+    function _acceptedTokenAmountFor(
+        uint256 _projectId,
+        address _token,
+        uint256 _amount,
+        bytes calldata _metadata
+    ) internal returns (uint256) {
+        // Make sure the project has set an accounting context for the token being paid.
+        if (_accountingContextForTokenOf[_projectId][_token].token == address(0)) {
+            revert TOKEN_NOT_ACCEPTED();
+        }
+
+        // If the terminal's token is ETH, override `_amount` with msg.value.
+        if (_token == JBTokens.ETH) return msg.value;
+
+        // Amount must be greater than 0.
+        if (msg.value != 0) revert NO_MSG_VALUE_ALLOWED();
+
+        // If the terminal is rerouting the tokens within its own functions, there's nothing to transfer.
+        if (msg.sender == address(this)) return _amount;
+
+        // Keep a reference to the allowance context parsed from the metadata.
+        JBSingleAllowanceData memory _allowance;
+
+        bool _quoteExists;
+        bytes memory _parsedMetadata;
+
+        // Unpack the allowance to use, if any, given by the frontend.
+        (_quoteExists, _parsedMetadata) =
+            JBDelegateMetadataLib.getMetadata(bytes4(uint32(uint160(address(this)))), _metadata);
+        if (_quoteExists) {
+            (_allowance) = abi.decode(_parsedMetadata, (JBSingleAllowanceData));
+        }
+
+        // Set the allowance to `spend` tokens for the user if needed.
+        if (_allowance.amount > 0) _permitAllowance(_allowance, _token);
```

Comment:
Might be missing a `PERMIT_ALLOWANCE_NOT_ENOUGH` check?

Reply:
won't it fail on transfer?

Reply:
add it in if you see fit! go ahead and get permit2 stuff to where it needs to be. plz also consider that `payer` param being sent to _pay.